DIPs/DIP1005.md

```diff
@@ -0,0 +1,781 @@
+    hasIndirections!(ElementEncodingType!T))
+{
+    ...
+}
+```
+
+Alternatively, the declaration may specify the exact symbols needed by using multiple imports:
+
+```d
+with (import std.meta : allSatisfy)
+with (import std.range : ElementEncodingType)
+with (import std.traits : hasIndirections, isDynamicArray, isIntegral)
+auto uninitializedArray(T, I...)(I sizes) nothrow @system
+if (isDynamicArray!T && allSatisfy!(isIntegral, I) &&
+    hasIndirections!(ElementEncodingType!T))
+{
+    ...
+}
+```
+
+#### Alternative: Lazy Imports
+
+Assume all `import`s are lazy without any change in the language. (This has already been implemented in the SDC compiler.) The way the scheme works, all `import`s seen are not yet executed but instead saved in a list of package/module names. Following that, the actual imports are triggered by one of two situations.
+
+First, consider the current module looks up a fully specified name:
+
+```d
+import module_a, module_b;
+void fun(T)(T value) if (module_a.condition!T)
+{
+    return module_b.process(value);
+}
+void fun(T)(T value) if (is(T == int)) { ... }
+```
+
+In this situation:
+
+* If `fun` is never looked up, neither `module_a` nor `module_b` needs to be loaded.
+* If `fun(42)` is used, even though the second overload is a match, then `module_a` must be loaded in order to ensure that `module_a.condition!int` is false so as to avoid ambiguity.
+* If `fun` is called with a non-`int` value, `module_a` is loaded to evaluate the template constraint. If the constraint is true, then `module_b` is also loaded so as to look up `process`.
+
+Let us note that full specification of symbols used may be enabled with ease by using the `static import` feature. We will henceforth refer to this setup as "the `static import` setup".
+
+Second, consider the situation (arguably more frequent in today's D code) when the current module does not fully specify names used. Instead, it imports the appropriate modules and relies on lookup to resolve symbols appropriately:
+
+```d
+import module_a, module_b;
+void fun(T)(T value) if (condition!T)
+{
+    return process(value);
+}
+void fun(T)(T value) if (is(T == int)) { ... }
+```
+
+In this situation:
+
+* If `fun` is never used, neither module must be loaded.
+* If `fun` is looked up, it will trigger an unspecified lookup for `condition`. This will trigger loading of both `module_a` and `module_b` (and generally all imports in the current module) so as to look up `condition` and ensure no ambiguity.
+
+The same applies to the setup in which `condition` is imported selectively from `module_a` but `module_b` is entirely imported:
+
+```d
+import module_a : condition;
+import module_b;
+void fun(T)(T value) if (condition!T)
+{
+    return process(value);
+}
+void fun(T)(T value) if (is(T == int)) { ... }
+```
+
+In this case, module `module_b` still needs to be opened if `fun` is looked up to ensure no ambiguity exists for `condition`.
```

Comment:
Wrong! condition is an alias on module level and does not overload with symbols behind `import module_b;`.